### PR TITLE
Allow retry just when there is a 409 response, issue #32

### PR DESCRIPTION
**Pull request by ccsalazarr-commercetools**

### CODE REVIEW

File: commercetools-convenience/src/main/java/io/sphere/sdk/sequencegenerators/CustomObjectBigIntegerNumberGenerator.java [hunk 1]
```diff
@@ -53,14 +54,23 @@ private CustomObjectBigIntegerNumberGenerator(final CustomObjectBigIntegerNumber
...
             return CompletableFutureUtils.recoverWith(bigIntegerCompletionStage, (error) -> tryGetNextNumber(timeToLive - 1, error));
         } else {
             return CompletableFutureUtils.failed(throwable);
         }
     }
 
+    private boolean isNullOrConcurrentException( final Throwable throwable ) {
+        if ( throwable == null ){
+            return true;
+        }else if(throwable.getCause() instanceof ConcurrentModificationException) {
```
Comment (yanns) [inline commercetools-convenience/src/main/java/io/sphere/sdk/sequencegenerators/CustomObjectBigIntegerNumberGenerator.java]: try to stay to the current code style:
```
    if (throwable == null) {
       ...
    } else if (...) {
    }
```

**Comment by schleichardt** [comment]
Tests are missing.

**Comment by ccsalazarr-commercetools** [comment]
@yanns and @schleichardt could you review the test (of the concurrent modification exception) to see if the approach is right, or if I have to do another kind of test.

**Comment by yanns** [comment]
for me, it's ok. Let's wait for @schleichardt here.

File: commercetools-convenience/src/test/java/io/sphere/sdk/sequencegenerators/BigIntegerNumberGeneratorIntegrationTest.java [hunk 2]
```diff
@@ -92,6 +100,44 @@ public void concurrentUsageTest() throws Exception {
...
+    public void failWithNotConcurrentModificationException() {
+        final String container = randomKey();
+        final String key = randomKey();
+        final CustomObjectBigIntegerNumberGeneratorConfig config =
+                CustomObjectBigIntegerNumberGeneratorConfigBuilder.of(getSphereClient(getBadRequestHttpClient()), key)
+                        .container(container)
+                        .build();
+        final BigIntegerNumberGenerator generator = CustomObjectBigIntegerNumberGenerator.of(config);
+        try {
+            generator.getNextNumber().toCompletableFuture().join();
```
Comment (schleichardt) [inline commercetools-convenience/src/test/java/io/sphere/sdk/sequencegenerators/BigIntegerNumberGeneratorIntegrationTest.java]: If this does not throw an exception then the test would be still green.

Have a look at http://joel-costigliola.github.io/assertj/assertj-core-features-highlight.html#exception-assertion and use `catchThrowable`.

Comment (schleichardt) [inline commercetools-convenience/src/test/java/io/sphere/sdk/sequencegenerators/BigIntegerNumberGeneratorIntegrationTest.java]: The test does not validate that it does not retry, it still could retry a lot of times and finally failes with the exception.

File: commercetools-convenience/src/test/java/io/sphere/sdk/sequencegenerators/BigIntegerNumberGeneratorIntegrationTest.java [hunk 3]
```diff
@@ -82,7 +90,7 @@ public void customObjectContainerAndKeyCanBeGiven() throws Exception {
     @Test
     public void concurrentUsageTest() throws Exception {
         final int firstNumber = 1;
-        final int lastNumber = 20;
+        final int lastNumber = 50;
```
Comment (schleichardt) [inline commercetools-convenience/src/test/java/io/sphere/sdk/sequencegenerators/BigIntegerNumberGeneratorIntegrationTest.java]: Why did you change that?

File: commercetools-convenience/src/test/java/io/sphere/sdk/sequencegenerators/BigIntegerNumberGeneratorIntegrationTest.java [hunk 4]
```diff
@@ -92,6 +101,44 @@ public void concurrentUsageTest() throws Exception {
...
+    public void failWithNotConcurrentModificationException() {
+        final String container = randomKey();
+        final String key = randomKey();
+        final CustomObjectBigIntegerNumberGeneratorConfig config =
+                CustomObjectBigIntegerNumberGeneratorConfigBuilder.of(getSphereClient(getBadRequestHttpClient()), key)
+                        .container(container)
+                        .build();
+        final BigIntegerNumberGenerator generator = CustomObjectBigIntegerNumberGenerator.of(config);
+
+        Throwable thrown = catchThrowable(() -> { generator.getNextNumber().toCompletableFuture().join(); });
```
Comment (schleichardt) [inline commercetools-convenience/src/test/java/io/sphere/sdk/sequencegenerators/BigIntegerNumberGeneratorIntegrationTest.java]: this should also work: `catchThrowable(() -> generator.getNextNumber().toCompletableFuture().join());`
